### PR TITLE
[secure-net] Improve the reliability of safety rules as a service

### PR DESCRIPTION
**Pull request by davidiw**
This PR does a handful of actions:
- Make the NetworkServer and hence SafetyRules attempt to acquire a new connection if there is an internal error (i.e., consensus has disconnected).
- Eliminate a few unwraps and unnecessary checks in consensus that involve SafetyRules
- Make the NetworkClient and hence the SafetyRules client attempt to acquire a new connection if there is an internal error (i.e., safety rules has disconnected)

Integration tests cover the code well. I'm planning on another PR to handle configuration and evaluating as much as possible locally in docker.

### CODE REVIEW

File: secure/net/src/lib.rs [hunk 1]
```diff
@@ -7,64 +7,95 @@
...
+//! error to let the caller know of the event. A follow up call will result in the service
+//! attempting to either reconnect in the case of a client or accept a new client in the case of a
+//! server.
 //!
 //! Internally both the client and server leverage a NetworkStream that communications in blocks
 //! where a block is a length prefixed array of bytes.
 
 use anyhow::{anyhow, ensure, Result};
+//#[cfg(test)]
+//use nix::sys::socket::{self, sockopt::ReusePort};
```
Comment (davidiw) [inline secure/net/src/lib.rs]: crap

File: secure/net/src/lib.rs [hunk 2]
```diff
@@ -204,4 +237,26 @@ mod test {
         let result = client.read().unwrap();
         assert_eq!(data, result);
     }
+
+    #[test]
+    fn test_shutdown() {
```
Comment (zekun000) [inline secure/net/src/lib.rs]: test server shutdown case too?

File: consensus/safety-rules/src/remote_service.rs [hunk 3]
```diff
@@ -30,12 +30,23 @@ pub fn execute<T: Payload>(
     let mut network_server = NetworkServer::new(listen_addr);
 
     loop {
-        let request = network_server.read().unwrap();
-        let response = serializer_service.handle_message(request).unwrap();
-        network_server.write(&response).unwrap();
+        match process_one_message(&mut network_server, &mut serializer_service) {
```
Comment (zekun000) [inline consensus/safety-rules/src/remote_service.rs]: if let Err(e) =

File: consensus/src/chained_bft/event_processor.rs [hunk 4]
```diff
@@ -516,38 +497,11 @@ impl<T: Payload> EventProcessor<T> {
...
-        debug_checked_verify_eq!(
-            self.safety_rules
-                .consensus_state()
-                .unwrap()
-                .last_voted_round(),
-            proposal_round
-        );
-        // Safety invariant: qc_parent <-- qc
-        // the preferred block round must be at least as large as qc_parent's round.
-        debug_checked_verify!(
```
Comment (zekun000) [inline consensus/src/chained_bft/event_processor.rs]: the intern would be sad to see this

File: consensus/src/chained_bft/event_processor.rs [hunk 5]
```diff
@@ -372,15 +368,17 @@ impl<T: Payload> EventProcessor<T> {
             // The timeout event is late: the node has already moved to another round.
             return;
         }
-        let last_voted_round = self
-            .safety_rules
-            .consensus_state()
-            .unwrap()
-            .last_voted_round();
+
+        let use_last_vote = if let Some((_, last_voted_round)) = self.last_vote_sent {
```
Comment (zekun000) [inline consensus/src/chained_bft/event_processor.rs]: `last_vote_sent` doesn't seem always consistent with safety_rules, but looks fine

File: secure/net/src/lib.rs [hunk 6]
```diff
@@ -23,55 +23,80 @@ use std::{
...
     /// Shutdown the internal network stream
-    pub fn shutdown(&self) -> Result<()> {
-        self.stream.shutdown()
+    pub fn shutdown(&mut self) -> Result<()> {
+        let stream = self
+            .stream
+            .take()
+            .ok_or_else(|| anyhow!("No active stream"))?;
+        stream.shutdown()?;
+        drop(stream);
```
Comment (zekun000) [inline secure/net/src/lib.rs]: why you need explicit drop?

File: secure/net/src/lib.rs [hunk 7]
```diff
@@ -239,24 +276,49 @@ mod test {
...
         assert!(server.read().is_err());
 
         let data = vec![4, 5, 6, 7];
         client.write(&data).unwrap();
         let result = server.read().unwrap();
         assert_eq!(data, result);
     }
+
+    #[test]
+    fn test_server_shutdown() {
```
Comment (zekun000) [inline secure/net/src/lib.rs]: isn't this commit belong closer to the first one?

File: consensus/src/chained_bft/event_processor.rs (hunk 4 above)
Comment (davidiw) [reply consensus/src/chained_bft/event_processor.rs]: Hahaha, spoke offline with @dmitri-perelman, we agreed it wasn't adding much value. Then if you take a look, I think these checks became irrelevant as we eliminated extra pieces of data within the block late last year.

File: secure/net/src/lib.rs (hunk 6 above)
Comment (davidiw) [reply secure/net/src/lib.rs]: Unlikely not necessary but it frees the port for reuse. I wanted to be consistent with the design on the server side.

File: secure/net/src/lib.rs (hunk 7 above)
Comment (davidiw) [reply secure/net/src/lib.rs]: I first verified I could perform restarts with the client in the earlier commits. Then I verified I could handle restarts with the server. Thus I separated them into separate commits.

File: consensus/safety-rules/src/remote_service.rs (hunk 3 above)
Comment (davidiw) [reply consensus/safety-rules/src/remote_service.rs]: wonder why clippy didn't complain :O

File: secure/net/src/lib.rs (hunk 6 above)
Comment (davidiw) [reply secure/net/src/lib.rs]: Turns out the explicit take action was strong enough to enforce without needing the drop. Seems shadowing could have resulted in this being the case. The way this is structured eliminates concerns about shadowing etc.

**Comment by davidiw** [comment]
> the network code doesn't look very rusty, but let's move forward

Looking forward to understanding why :P
I suspect it is a bit unnatural in that other languages likely would make it obvious 
* how to cleanly shutdown (shutdown with no need for explicitly dropping, etc)
* recognize that a remote side has been closed (returning an error instead of 0 bytes)

There is some opportunity to ask the question what is the right location to handle reconnect logic, I reckon since NetworkClient / NetworkServer are intended for use in our applications, we can tailor them to our needs. Namely that they never really die.

thanks for accepting :)

@bors-libra r=zekun000

**Pull request closed without merging**